### PR TITLE
Good to ping

### PR DESCRIPTION
**Pull request by NamsooCho**
Add function to check if we need to send ping.
function name is `is_redundant_ping`

### CODE REVIEW

File: src/toxcore/onion/request_queue.rs [hunk 1]
```diff
@@ -6,7 +6,7 @@ use crate::toxcore::time::*;
 
 #[derive(Clone, Debug)]
 pub struct RequestQueue<T> {
-    ping_map: HashMap<u64, (Instant, T)>,
+    pub ping_map: HashMap<u64, (Instant, T)>,
```
Comment (kurnevsky) [inline src/toxcore/onion/request_queue.rs]: Add method that will return optional timestamp rather than making it public.

File: src/toxcore/onion/onion_client.rs [hunk 2]
```diff
@@ -292,6 +292,25 @@ impl OnionClient {
...
+                if request_data.pk == pk {
+                    Some(time)
+                } else {
+                    None
+                }
+            })
+            .filter(|time| clock_elapsed(**time) < Duration::from_secs(ONION_NODE_PING_INTERVAL))
+            .collect::<Vec<_>>();
+
+        if last_ping_time.is_empty() {
```
Comment (kurnevsky) [inline src/toxcore/onion/onion_client.rs]: !last_ping_time.is_empty()

Comment (kurnevsky) [inline src/toxcore/onion/onion_client.rs]: Change it to `map_or` after method adding.

File: src/toxcore/onion/request_queue.rs [hunk 3]
```diff
@@ -50,4 +50,28 @@ impl<T> RequestQueue<T> { // TODO: unify with DHT?
...
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let len = ping_times.len();
+        if len == 0 {
+            None
+        } else if len > 1 {
+            warn!("There are multiple same pings to a peer");
+            None
```
Comment (kurnevsky) [inline src/toxcore/onion/request_queue.rs]: Do we need this warn? Why don't just use `find_map`? Also returning `None` here will lead to more duplications.

File: src/toxcore/onion/request_queue.rs [hunk 4]
```diff
@@ -50,4 +50,28 @@ impl<T> RequestQueue<T> { // TODO: unify with DHT?
             clock_elapsed(time) <= timeout
         );
     }
+
+    /// Returns the Instant of a PublicKey, used to check if the PK is timedout
+    pub fn get_ping_time<F>(&self, f: F) -> Option<&Instant>
```
Comment (kurnevsky) [inline src/toxcore/onion/request_queue.rs]: Maybe make it more generic, return `&(Instant, T)` and call it just `find`?

File: src/toxcore/onion/onion_client.rs [hunk 5]
```diff
@@ -292,6 +292,22 @@ impl OnionClient {
...
+        let check_pks = |data: &AnnounceRequestData| -> bool {
+            if let Some(friend_pk) = data.friend_pk {
+                data.pk == pk && search_pk == friend_pk
+            } else {
+                data.pk == pk
+            }
+        };
+        let ping_time = request_queue.get_ping_time(check_pks);
+        if let Some(ping_time) = ping_time {
+            clock_elapsed(*ping_time) < Duration::from_secs(ONION_NODE_PING_INTERVAL)
```
Comment (kurnevsky) [inline src/toxcore/onion/onion_client.rs]: It should be `MIN_NODE_PING_TIME = 10`.

Comment (NamsooCho) [reply src/toxcore/onion/onion_client.rs]: done

File: src/toxcore/onion/request_queue.rs [hunk 6]
```diff
@@ -50,4 +50,17 @@ impl<T> RequestQueue<T> { // TODO: unify with DHT?
             clock_elapsed(time) <= timeout
         );
     }
+
+    /// Returns the Instant of a HashMap according to `f`.
+    pub fn find<F>(&self, f: F) -> Option<(&Instant, &T)>
+        where F: Fn(&T) -> bool {
+        self.ping_map.iter()
+            .find_map(|(_ping_id, (ping_time, data))| {
```
Comment (kurnevsky) [inline src/toxcore/onion/request_queue.rs]: Use `ping_map.values` since you don't need `ping_id` here.

File: src/toxcore/onion/onion_client.rs [hunk 7]
```diff
@@ -292,6 +294,22 @@ impl OnionClient {
         }
     }
 
+    fn is_redundant_ping(pk: PublicKey, search_pk: PublicKey, request_queue: &RequestQueue<AnnounceRequestData>) -> bool {
+        let check_pks = |data: &AnnounceRequestData| -> bool {
+            if let Some(friend_pk) = data.friend_pk {
+                data.pk == pk && search_pk == friend_pk
+            } else {
+                data.pk == pk
```
Comment (kurnevsky) [inline src/toxcore/onion/onion_client.rs]: Here you should check that `search_pk` equals our `dht_pk`, otherwise you can find previous announce request for current search request. Then it can be rewritten to:
```rust
let request_search_pk = if Some(friend_pk) = data.friend_pk {
    friend_pk
} else {
    dht_pk
};
data.pk == pk && search_pk == request_search_pk
```
I think it looks better since here you just check that npde's pk and search pk from new request equals to same pks from old request.

Comment (NamsooCho) [reply src/toxcore/onion/onion_client.rs]: done

File: src/toxcore/onion/onion_client.rs [hunk 8]
```diff
@@ -358,12 +377,15 @@ impl OnionClient {
...
 
         for node in &payload.nodes {
-            // TODO: reduce pings frequency? (see good_to_ping in c-toxcore)
-
             if !nodes_list.can_add(&announce_packet_data.search_pk, &node, /* evict */ true) {
                 continue;
             }
 
+            // To prevent to send redundant ping packet.
+            if self.is_redundant_ping(node.pk, announce_packet_data.search_pk, &state.announce_requests) {
```
Comment (kpp) [inline src/toxcore/onion/onion_client.rs]: Why did you add self as the first arg? It was not here before. Try to call clippy  on the project

Comment (NamsooCho) [reply src/toxcore/onion/onion_client.rs]: https://github.com/tox-rs/tox/blob/00a77960257b865aa29fc4e662817d9021498a68/src/toxcore/onion/onion_client.rs#L302

Now self is needed.

**Comment by NamsooCho** [comment]
>How about .filter instead of find_map
done

File: src/toxcore/onion/request_queue.rs [hunk 9]
```diff
@@ -50,4 +50,14 @@ impl<T> RequestQueue<T> { // TODO: unify with DHT?
...
     }
+
+    /// Returns the Instant of a HashMap according to `f`.
+    pub fn find<F>(&self, f: F) -> Option<&(Instant, T)>
+        where F: Fn(&T) -> bool {
+        self.ping_map.values()
+            .filter(|(_ping_time, data)| {
+                f(data)
+            })
+            .next()
```
Comment (kurnevsky) [inline src/toxcore/onion/request_queue.rs]: Then it can be just `find` :)

Comment (NamsooCho) [reply src/toxcore/onion/request_queue.rs]: done

File: src/toxcore/onion/onion_client.rs (hunk 8 above)
Comment (kpp) [reply src/toxcore/onion/onion_client.rs]: OK

File: src/toxcore/onion/request_queue.rs [hunk 10]
```diff
@@ -50,4 +50,11 @@ impl<T> RequestQueue<T> { // TODO: unify with DHT?
             clock_elapsed(time) <= timeout
         );
     }
+
+    /// Returns the Instant of a HashMap according to `f`.
+    pub fn find<F>(&self, f: F) -> Option<&(Instant, T)>
```
Comment (kpp) [inline src/toxcore/onion/request_queue.rs]: find_by_value

File: src/toxcore/onion/onion_client.rs [hunk 11]
```diff
@@ -292,6 +294,25 @@ impl OnionClient {
...
+        let check_pks = |data: &AnnounceRequestData| -> bool {
+            let request_search_pk = if let Some(friend_pk) = data.friend_pk {
+                friend_pk
+            } else {
+                self.dht.pk
+            };
+            data.pk == pk && search_pk == request_search_pk
+        };
+        let request = request_queue.get_values()
+            .filter(|(ping_time, _request_data)| clock_elapsed(*ping_time) <= request_queue.get_timeout())
```
Comment (kurnevsky) [inline src/toxcore/onion/onion_client.rs]: Wrong place to filter timed out records. It doesn't make sense just because next you check time anyway. It rather should be [here](https://github.com/tox-rs/tox/pull/325/files#diff-6e41041a9b7f108d886020985a2290c2R57) to make this function consistent regardless of `clear_timed_out` execution.

Comment (kpp) [reply src/toxcore/onion/onion_client.rs]: Why?

File: src/toxcore/onion/request_queue.rs [hunk 12]
```diff
@@ -50,4 +50,16 @@ impl<T> RequestQueue<T> { // TODO: unify with DHT?
             clock_elapsed(time) <= timeout
         );
     }
+
+    /// Returns the Values of ping_map.
+    pub fn get_values(&self) -> impl Iterator<Item=(Instant, &T)> {
+        self.ping_map
+            .values()
```
Comment (kpp) [inline src/toxcore/onion/request_queue.rs]: Namsoo, add `.filter(|(ping_time, _request_data)| clock_elapsed(*ping_time) <= request_queue.get_timeout())` here, and add mark to doc comment that we return fresh values

Comment (NamsooCho) [reply src/toxcore/onion/request_queue.rs]: done

File: src/toxcore/onion/onion_client.rs [hunk 13]
```diff
@@ -292,6 +294,24 @@ impl OnionClient {
...
+            let request_search_pk = if let Some(friend_pk) = data.friend_pk {
+                friend_pk
+            } else {
+                self.dht.pk
+            };
+            data.pk == pk && search_pk == request_search_pk
+        };
+        let request = request_queue.get_values()
+            .find(|(_ping_time, request_data)| check_pks(request_data));
+        if let Some((ping_time, _request_data)) = request {
```
Comment (kpp) [inline src/toxcore/onion/onion_client.rs]: Now the check for timeout is redundant here, check with is_some

Comment (NamsooCho) [reply src/toxcore/onion/onion_client.rs]: As I know, the timeout of RequestQueue is the timeout of Queue itself and this check is for preventing redundant ping.

Comment (NamsooCho) [reply src/toxcore/onion/onion_client.rs]: RequestQueue timeout is ANNOUNCE_TIMEOUT which is 300 seconds and cheking timeout is 10 seconds.

File: src/toxcore/onion/onion_client.rs [hunk 14]
```diff
@@ -292,6 +294,24 @@ impl OnionClient {
...
+        let check_pks = |data: &AnnounceRequestData| -> bool {
+            let request_search_pk = if let Some(friend_pk) = data.friend_pk {
+                friend_pk
+            } else {
+                self.dht.pk
+            };
+            data.pk == pk && search_pk == request_search_pk
+        };
+        let request = request_queue.get_values()
+            .find(|(_ping_time, request_data)| check_pks(request_data));
```
Comment (kurnevsky) [inline src/toxcore/onion/onion_client.rs]: After yesterday's discussion with @kpp I realized that this is not a correct check. We have different timeouts for allowing new requests and for expiration of old requests. So it's possible that we have different entries in the queue for the same request. Which means instead of finding the first one we should check all matches, i.e. instead of `find` we should use `any` or `all`. Otherwise we can send a new request even though we sent it recently.

Comment (NamsooCho) [reply src/toxcore/onion/onion_client.rs]: done
